### PR TITLE
Separate python deps

### PR DESCRIPTION
**Pull request by browningjp-datatonic**
# Description

The aim of this PR is to separate the Python dependencies for each pipeline component from each other component and from the ML pipelines themselves. There are a lot of files changed (!) but this mostly consists of files moving and some import statements being updated. The key changes are:

- Pipeline components have been moved out of the `pipelines/kfp_components` directory and into `pipeline_components`
- Each group of pipeline components (e.g. `aiplatform`, `bigquery`, `helpers` etc) has its own Python virtual environment managed with `pipenv` and its own tests in its directory
- The existing `pipelines` and `tests` directories are moved to a new top-level directory called `pipelines`, which also now contains the `Pipfile` and `Pipfile.lock` for the ML pipelines virtual env (previously in the root of the repo)
- Instead of importing the `@component` functions directly for use in ML pipelines, the pipeline components are first compiled to YAML (in their respective virtual environment), before being imported from YAML for use in the ML pipelines
- Updates to python imports, Makefile, CI/CD pipelines and docs related to the above changes

# How has this been tested?

- Passing automated tests set up on my fork of the repo - see equivalent PR [here](https://github.com/browningjp-datatonic/vertex-pipelines-end-to-end-samples/pull/2)
  - The xgboost E2E test is still failing due to #26 - this PR will enable us to fix this more easily by modifying the dependencies for (only) the XGBoost components
- Make commands tested locally

# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have successfully run the E2E tests
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~
- [x] New and existing unit tests pass locally with my changes
- [x] I have updated any relevant documentation to reflect my changes
- [x] I have assigned a reviewer and messaged them

### CODE REVIEW

**Comment by simonbutt-datatonic** [comment]
Add:
```
pipeline_components/evaluation/example_count.py	pipeline_components/evaluation/plots_num_feat.html
pipeline_components/evaluation/plots_overall.html
```
to .gitignore or clean during specific test. These files are created when running `make test-all-components`

File: pipeline_components/_tensorflow/_tensorflow/train/component.yaml [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+
+      if ! [ -x "$(command -v pip)" ]; then
+          python3 -m ensurepip || python3 -m ensurepip --user || apt-get install python3-pip
+      fi
+
+      PIP_DISABLE_PIP_VERSION_CHECK=1 python3 -m pip install --quiet     --no-warn-script-location 'tensorflow==2.7.1' 'kfp==1.8.9' && "$0" "$@"
+    - sh
+    - -ec
+    - |
+      program_path=$(mktemp -d)
```
Comment (simonbutt-datatonic) [inline pipeline_components/_tensorflow/_tensorflow/train/component.yaml:85]: This much code shouldn't be in yaml. It makes it very hard to CR, test or develop on. Same for all `component.yaml` files

File: pipeline_components/_tfdv/_tfdv/show_anomalies/component.yaml [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      fi
+
+      PIP_DISABLE_PIP_VERSION_CHECK=1 python3 -m pip install --quiet     --no-warn-script-location 'tensorflow-data-validation==1.6.0' 'kfp==1.8.9' && "$0" "$@"
+    - sh
+    - -ec
+    - |
+      program_path=$(mktemp -d)
+      printf "%s" "$0" > "$program_path/ephemeral_component.py"
+      python3 -m kfp.v2.components.executor_main                         --component_module_path                         "$program_path/ephemeral_component.py"                         "$@"
+    - "\nimport kfp\nfrom kfp.v2 import dsl\nfrom kfp.v2.dsl import *\nfrom typing\
```
Comment (simonbutt-datatonic) [inline pipeline_components/_tfdv/_tfdv/show_anomalies/component.yaml:47]: Is there a better way to store this for readability?

**Comment by simonbutt-datatonic** [comment]
`pipeline_components/_tfdv/_tfdv/conftest.py` if empty, is it needed?

File: pipeline_components/_tfdv/_tfdv/show_anomalies/component.yaml (hunk 2 above)
Comment (browningjp-datatonic) [reply pipeline_components/_tfdv/_tfdv/show_anomalies/component.yaml]: remove component.yaml files

File: pipeline_components/_tensorflow/_tensorflow/train/component.yaml (hunk 1 above)
Comment (browningjp-datatonic) [reply pipeline_components/_tensorflow/_tensorflow/train/component.yaml]: removed component.yaml files

**Comment by browningjp-datatonic** [comment]
> Add:
> 
> ```
> pipeline_components/evaluation/example_count.py 
> pipeline_components/evaluation/plots_num_feat.html
> pipeline_components/evaluation/plots_overall.html
> ```
> 
> to .gitignore or clean during specific test. These files are created when running `make test-all-components`

Added to .gitignore for this PR. Created #29 to track it and will address properly in a future PR

**Comment by browningjp-datatonic** [comment]
> `pipeline_components/_tfdv/_tfdv/conftest.py` if empty, is it needed?

Yes this is required for pytest to pick up the directory so that the contents can be imported during tests

**Comment by simonbutt-datatonic** [comment]
> 

In that case potentially add a comment stating the purpose in the file. Otherwise can image people developing on the template could remove without realising the failing test consequence

**Review by simonbutt-datatonic — APPROVED** [review]
Small comment related to `_tfdv/conftest.py`, otherwise LGTM.
Nice work and overall great change to the codebase.

**Comment by browningjp-datatonic** [comment]
> > 
> 
> In that case potentially add a comment stating the purpose in the file. Otherwise can image people developing on the template could remove without realising the failing test consequence

Added

**Comment by browningjp-datatonic** [comment]
@Linchin ready for your review :)

**Pull request merged**